### PR TITLE
worked on the post , getall and delete endpoint

### PR DESCRIPTION
**Pull request by hackercaleb**

### CODE REVIEW

File: src/controllers/shortener.controller.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import { Request, Response } from 'express';
+import mongoose from 'mongoose';
+import { customAlphabet } from 'nanoid';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]: Remove this eslint-disable

File: src/controllers/shortener.controller.ts
Comment (greatertomi) [inline src/controllers/shortener.controller.ts:1]: Delete all the console.log in this file

File: src/controllers/shortener.controller.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import ShortenedURL, { IShortenedURL } from '../models/shortener.model';
+
+// Create a custom nanoid generator for generating short URLs
+const nanoidGenerator = customAlphabet(
+  '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz',
+  5
+);
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]: Extract this to a custom function and export it.

File: .env [hunk 3]
```diff
@@ -0,0 +1,4 @@
+MONGO_USERNAME=caleboshalusi
+MONGO_PASSWORD=4UErhUhlMqFGRY4k
+MONGO_CLUSTER_NAME=cluster0
+MONGO_URL = mongodb+srv://caleboshalusi:XQskTG1hGgs7VuGn@cluster0.dqdxx7i.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0
```
Comment (greatertomi) [inline .env]: Remove this .env file and create an .env.example file

File: src/controllers/shortener.controller.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+interface CreateUrlBody {
+  originalUrl: string;
+  customName?: string;
+}
+
+export const createUrls = async (req: Request, res: Response): Promise<void> => {
+  const { originalUrl, customName } = req.body as CreateUrlBody;
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]: - The requirement expects you to receive url

```suggestion
  const { url, customName } = req.body as CreateUrlBody;
```

File: src/controllers/shortener.controller.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    if (existingUrl) {
+      res.json({
+        message: 'Original url already exists',
+        data: { shortUrl: existingUrl.shortUrl }
+      });
+      return;
+    }
+
+    const shortUrl = customName || nanoidGenerator();
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]: > If the custom name has spaces, replace the spaces with dashes before adding it to the url. Save the custom name with spaces into the database.

Using the `customName` like this allows for the customName to have space

File: src/controllers/shortener.controller.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    });
+  } catch (error) {
+    console.error('Error:', error);
+    res.status(500).json({ message: 'Server error' });
+  }
+};
+
+export const getUrls = async (req: Request, res: Response) => {
+  try {
+    const getAllUrls = await ShortenedURL.find({});
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]:
```suggestion
    const getAllUrls = await ShortenedURL.find();
```

File: src/controllers/shortener.controller.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    // if there are no url in the database return empty array
+    if (!getAllUrls || getAllUrls.length === 0) {
+      return res.status(200).json([]);
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]: Extract all the HTTP_CODES in this PR to a constant file and use them from there

File: src/controllers/shortener.controller.ts [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    // to get all urls
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]: This comment is not necessary
```suggestion
```

File: src/routes/shortener.route.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import { Router } from 'express';
+import { createUrls, getUrls, deleteUrl } from '../controllers/shortener.controller';
+//import {
+// urlValidatorRules,
+// updateUrlValidatorRules,
+//} from '../middleware/urlValidator';
+//import { validate } from '../middleware/validate';
```
Comment (greatertomi) [inline src/routes/shortener.route.ts]: Delete this

File: src/models/shortener.model.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    unique: true
+  },
+  originalUrl: {
+    type: String,
+    required: true
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  }
```
Comment (greatertomi) [inline src/models/shortener.model.ts:28]: When fetching all the urls, you can improve the model by changing the `_id` to `id`. Remove the underscore. Also delete the __v

Something like this below should help.
Screenshot 2024-04-06 at 13 33 24

File: src/controllers/shortener.controller.ts [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+    if (!originalUrlFind) {
+      return res.status(404).json({ error: 'Original URL not found' });
+    }
+
+    // Remove the document from the database
+    await ShortenedURL.findByIdAndDelete(objectId);
+
+    return res.status(200).json({
+      message: 'URL deleted successfully',
+      data: null
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]: When you delete a data, return the deleted data not `null`

File: src/controllers/shortener.controller.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    const existingUrl = await ShortenedURL.findOne({ originalUrl });
+    console.log('Existing URL:', existingUrl);
+
+    if (existingUrl) {
+      res.json({
+        message: 'Original url already exists',
+        data: { shortUrl: existingUrl.shortUrl }
+      });
+      return;
+    }
```
Comment (greatertomi) [inline src/controllers/shortener.controller.ts]:
```suggestion
    if (existingUrl) {
      return res.json({
        message: 'Original url already exists',
        data: { shortUrl: existingUrl.shortUrl }
      });
    
    }
```

File: src/index.ts [hunk 13]
```diff
@@ -14,3 +15,22 @@ app.get('/', (req: Request, res: Response) => {
...
+      throw new Error('MONGO_URL is not defined');
+    }
+
+    await mongoose.connect(mongoURL);
+
+    console.log('MongoDB Atlas Connected');
+  } catch (error) {
+    console.error('Error connecting to MongoDB Atlas:', error);
+  }
+};
```
Comment (greatertomi) [inline src/index.ts]: Extract this database connection logic to a separate file

**Review by greatertomi — CHANGES_REQUESTED** [review]
A good start. But there could be more improvements
- When I try to use a custom name more than once, it throws a server error. It should return an error message that custom name is already used
- You need to add validation to the post route. When I send an empty object, it throws an error.

File: src/controllers/shortener.controller.ts (hunk 1 above)
Comment (greatertomi) [reply src/controllers/shortener.controller.ts]: Fixed on 73c307e

**Comment by hackercaleb** [comment]
fixed on aea946a4e3e0e60e96df35aa5e2397e8cc24b0cf

**Review by greatertomi — APPROVED** [review]

**Pull request merged**